src/functions/InModuleScope.ps1

```diff
@@ -75,26 +75,48 @@
     )
 
     $module = Get-ScriptModule -ModuleName $ModuleName -ErrorAction Stop
+    $sessionState = Set-SessionStateHint -PassThru -Hint "Module - $($module.Name)" -SessionState $module.SessionState
 
-    # TODO: could this simply be $PSCmdlet.SessionState? Because the original scope we are moving from
-    # is the scope in which this command is running, right?
-    # $originalState = $Pester.SessionState
-    # $originalScriptBlockScope = Get-ScriptBlockScope -ScriptBlock $ScriptBlock
-
-    # try {
-    # $sessionState = Set-SessionStateHint -PassThru -Hint "Module - $($module.Name)" -SessionState $module.SessionState
-    # $Pester.SessionState = $sessionState
-    # Set-ScriptBlockScope -ScriptBlock $ScriptBlock -SessionState $sessionState
-
-    # do {
-    # Write-ScriptBlockInvocationHint -Hint "InModuleScope" -ScriptBlock $ScriptBlock
-    & $module $ScriptBlock @Parameters @ArgumentList
-    # } until ($true)
-    # }
-    # finally {
-    # $Pester.SessionState = $originalState
-    # Set-ScriptBlockScope -ScriptBlock $ScriptBlock -SessionStateInternal $originalScriptBlockScope
-    # }
+    $wrapper = {
+        param (
+            [Parameter(Mandatory = $true)]
+            [scriptblock]
+            ${Script Block},
+
+            [hashtable]
+            $___Parameters___ = @{ },
+
+            [object[]]
+            $___ArgumentList___ = @(),
+
+            [System.Management.Automation.SessionState]
+            ${Session State},
+
+            ${Set Dynamic Parameter Variable}
+        )
+
+        # This script block is used to create variables for provided parameters that
+        # the real scriptblock can inherit. Makes defining a param-block optional.
+
+        & ${Set Dynamic Parameter Variable} -SessionState ${Session State} -Parameters $___Parameters___
+
+        # define this in the current scope to be used instead of $PSBoundParameter if needed
+        # $PesterBoundParameters = if ($null -ne $___Parameters___) { $___Parameters___ } else { @{} }
+        & ${Script Block} @___Parameters___ @___ArgumentList___
```

Comment:
This is problematic when you don't define param and use $args. In that case parameters will become the first items in the $args array. You should do `@___ArgumentList___ @___Parameters___` instead that way it works slightly differently than normal powershell would, but your args will still work, and parameters will also bind if present.

Reply:
Good point, adding test to make sure first item in args is exptected argument. I'm checking if `$___Parameters___ ` is empty and avoid splatting it all together. That way args won't be polluted with a null-item at the end.